adapters/audienceNetwork/facebook.go

```diff
@@ -37,267 +41,416 @@ func (a *FacebookAdapter) Name() string {
 	return "audienceNetwork"
 }
 
-// Facebook likes to parallelize to minimize latency
-func (a *FacebookAdapter) SplitAdUnits() bool {
-	return true
-}
-
 func (a *FacebookAdapter) SkipNoCookies() bool {
 	return false
 }
 
-type facebookParams struct {
-	PlacementId string `json:"placementId"`
+type facebookReqExt struct {
+	PlatformId string `json:"platformid"`
+	AuthId     string `json:"authentication_id"`
 }
 
-func coinFlip() bool {
-	return rand.Intn(2) != 0
-}
+func (this *FacebookAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "no impressions provided",
+		}}
+	}
 
-func (a *FacebookAdapter) callOne(ctx context.Context, reqJSON bytes.Buffer) (result adapters.CallOneResult, err error) {
-	url := a.URI
-	if coinFlip() {
-		//50% of traffic to non-secure endpoint
-		url = a.nonSecureUri
+	if request.User == nil || request.User.BuyerUID == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Missing bidder token in 'user.buyeruid'",
+		}}
 	}
-	httpReq, _ := http.NewRequest("POST", url, &reqJSON)
-	httpReq.Header.Add("Content-Type", "application/json")
-	httpReq.Header.Add("Accept", "application/json")
-
-	anResp, e := ctxhttp.Do(ctx, a.http.Client, httpReq)
-	if e != nil {
-		err = e
-		return
+
+	reqs, errs := this.buildRequests(request)
+	return reqs, errs
+}
+
+func (this *FacebookAdapter) buildRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		panic("invalid number of impressions")
```

Comment:
I don't think the adapter should panic like this. It should return an error so that the error can be passed back in the response and the incorrect config/request can be fixed. A panic won't go back in the response, and can trigger our alerts that an adapter is panicing and needs to be fixed.

Reply:
Agreed, but this panic is there because this should never happen unless the code is wrong. I added a len(request.Imp) == 0 earlier in `MakeRequests` that returns an error, so the panic on L72 should never happen, its just there to safeguard from future iterations

Reply:
I see what you're saying. IMHO it would be easier to understand if you move the check which produces an error from `MakeRequests` right here, or omit the check here since it covered earlier by the only usage of this method. I don't think you need to `panic`. Future devs will understand and the panic is raising a few eyebrows.

Reply:
Thats fair, i'll remove it

Reply:
done